drivers/pwm/pwm_led_esp32.c

```diff
@@ -1,478 +1,351 @@
-#define TIMER_LS_BIT_NUM(i) (CONFIG_PWM_LED_ESP32_LS_TIMER ## i ## _BIT_NUM)
+static struct pwm_ledc_esp32_channel_config channel_config[] = {
+	DT_INST_FOREACH_CHILD(0, CHANNEL_CONFIG)
+};
 
-const static struct pwm_led_esp32_config pwm_led_esp32_config = {
-	.timer_cfg[PWM_LED_ESP32_HIGH_SPEED][0] = {
-		.bit_num = TIMER_HS_BIT_NUM(0),
-		.freq = TIMER_HS_FREQ(0),
-	},
-	.timer_cfg[PWM_LED_ESP32_HIGH_SPEED][1] = {
-		.bit_num = TIMER_HS_BIT_NUM(1),
-		.freq = TIMER_HS_FREQ(1),
-	},
-	.timer_cfg[PWM_LED_ESP32_HIGH_SPEED][2] = {
-		.bit_num = TIMER_HS_BIT_NUM(2),
-		.freq = TIMER_HS_FREQ(2),
-	},
-	.timer_cfg[PWM_LED_ESP32_HIGH_SPEED][2] = {
-		.bit_num = TIMER_HS_BIT_NUM(2),
-		.freq = TIMER_HS_FREQ(2),
-	},
-	.timer_cfg[PWM_LED_ESP32_HIGH_SPEED][3] = {
-		.bit_num = TIMER_HS_BIT_NUM(3),
-		.freq = TIMER_HS_FREQ(3),
-	},
-	.timer_cfg[PWM_LED_ESP32_LOW_SPEED][0] = {
-		.bit_num = TIMER_LS_BIT_NUM(0),
-		.freq = TIMER_LS_FREQ(0),
-	},
-	.timer_cfg[PWM_LED_ESP32_LOW_SPEED][1] = {
-		.bit_num = TIMER_LS_BIT_NUM(1),
-		.freq = TIMER_LS_FREQ(1),
-	},
-	.timer_cfg[PWM_LED_ESP32_LOW_SPEED][2] = {
-		.bit_num = TIMER_LS_BIT_NUM(2),
-		.freq = TIMER_LS_FREQ(2),
-	},
-	.timer_cfg[PWM_LED_ESP32_LOW_SPEED][3] = {
-		.bit_num = TIMER_LS_BIT_NUM(3),
-		.freq = TIMER_LS_FREQ(3),
-	},
-	.ch_cfg[0] = {
-		.timer = CH_HS_TIMER(0),
-		.gpio = CH_HS_GPIO(0),
-	},
-	.ch_cfg[1] = {
-		.timer = CH_HS_TIMER(1),
-		.gpio = CH_HS_GPIO(1),
-	},
-	.ch_cfg[2] = {
-		.timer = CH_HS_TIMER(2),
-		.gpio = CH_HS_GPIO(2),
-	},
-	.ch_cfg[3] = {
-		.timer = CH_HS_TIMER(3),
-		.gpio = CH_HS_GPIO(3),
-	},
-	.ch_cfg[4] = {
-		.timer = CH_HS_TIMER(4),
-		.gpio = CH_HS_GPIO(4),
-	},
-	.ch_cfg[5] = {
-		.timer = CH_HS_TIMER(5),
-		.gpio = CH_HS_GPIO(5),
-	},
-	.ch_cfg[6] = {
-		.timer = CH_HS_TIMER(6),
-		.gpio = CH_HS_GPIO(6),
-	},
-	.ch_cfg[7] = {
-		.timer = CH_HS_TIMER(7),
-		.gpio = CH_HS_GPIO(7),
-	},
-	.ch_cfg[8] = {
-		.timer = CH_LS_TIMER(0),
-		.gpio = CH_LS_GPIO(0),
-	},
-	.ch_cfg[9] = {
-		.timer = CH_LS_TIMER(1),
-		.gpio = CH_LS_GPIO(1),
-	},
-	.ch_cfg[10] = {
-		.timer = CH_LS_TIMER(2),
-		.gpio = CH_LS_GPIO(2),
-	},
-	.ch_cfg[11] = {
-		.timer = CH_LS_TIMER(3),
-		.gpio = CH_LS_GPIO(3),
-	},
-	.ch_cfg[12] = {
-		.timer = CH_LS_TIMER(4),
-		.gpio = CH_LS_GPIO(4),
-	},
-	.ch_cfg[13] = {
-		.timer = CH_LS_TIMER(5),
-		.gpio = CH_LS_GPIO(5),
-	},
-	.ch_cfg[14] = {
-		.timer = CH_LS_TIMER(6),
-		.gpio = CH_LS_GPIO(6),
-	},
-	.ch_cfg[15] = {
-		.timer = CH_LS_TIMER(7),
-		.gpio = CH_LS_GPIO(7),
+static struct pwm_ledc_esp32_config pwm_ledc_esp32_config = {
+	.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0),
+	.clock_dev = DEVICE_DT_GET(DT_INST_CLOCKS_CTLR(0)),
+	.clock_subsys = (clock_control_subsys_t)DT_INST_CLOCKS_CELL(0, offset),
+	.channel_config = channel_config,
+	.channel_len = ARRAY_SIZE(channel_config),
+};
+
+static struct pwm_ledc_esp32_data pwm_ledc_esp32_data = {
+	.hal = {
+		.dev = (ledc_dev_t *) DT_REG_ADDR(DT_NODELABEL(ledc0)),
```

Comment:
why isn't this using `_INST` macros?

Reply:
Do you mean like .dev = (ledc_dev_t *) DT_REG_ADDR(DT_INST(0, espressif_esp32_ledc)) ?